# THIS SKETCHPAD HAS BEEN CLEANED INTO 'SKETCHPAD2' 

In [1]:
cd /home/jupyter/PACE

/home/jupyter/PACE


In [2]:
ls

Declarations_19722_Merge.cpg  Declarations_19722_Merge.shp.xml
Declarations_19722_Merge.dbf  Declarations_19722_Merge.shx
Declarations_19722_Merge.prj  aoi_flood_footprint.geojson
Declarations_19722_Merge.sbn  data/
Declarations_19722_Merge.sbx  notebooks/
Declarations_19722_Merge.shp


In [3]:
import traceback
import warnings

from io import StringIO
import ee
import time
import os
from glob import glob
from typing import Optional, Callable, List, Tuple
from shapely.geometry import mapping, Polygon
import numpy as np
import geopandas as gpd
import pandas as pd
import fsspec
from datetime import datetime, timezone
import math
import json

from ml4floods.data import ee_download, utils


# Calling some functions

Removed the full function call and used the following instead.

In [4]:
get_count_function = ee_download.get_count_function

In [5]:
get_landsat_collection = ee_download.get_landsat_collection

In [6]:
collection_mosaic_day = ee_download.collection_mosaic_day

In [7]:
BANDS_NAMES = ee_download.BANDS_NAMES

Calling the aoi file

# Reading the geojson file

In [8]:
sydney_aoi= gpd.read_file('data/2022-07-21_aois.geojson')

In [9]:
sydney_aoi

aoi_code                                           geometry
0      AOI01  POLYGON ((147.09960 -31.24462, 147.30960 -31.2...
1      AOI01  POLYGON ((147.09960 -31.24462, 147.30960 -31.2...
2      AOI02  POLYGON ((147.09960 -31.04462, 147.30960 -31.0...
3      AOI02  POLYGON ((147.09960 -31.04462, 147.30960 -31.0...
4    AOI0289  POLYGON ((147.49960 -32.44461, 147.70960 -32.4...
..       ...                                                ...
476    AOI97  POLYGON ((150.29960 -34.24462, 150.50960 -34.2...
477    AOI98  POLYGON ((150.29960 -34.04462, 150.50960 -34.0...
478    AOI98  POLYGON ((150.29960 -34.04462, 150.50960 -34.0...
479    AOI99  POLYGON ((150.29960 -33.84462, 150.50960 -33.8...
480    AOI99  POLYGON ((150.29960 -33.84462, 150.50960 -33.8...

[481 rows x 2 columns]

In [10]:
sydney_aoi.geometry

0      POLYGON ((147.09960 -31.24462, 147.30960 -31.2...
1      POLYGON ((147.09960 -31.24462, 147.30960 -31.2...
2      POLYGON ((147.09960 -31.04462, 147.30960 -31.0...
3      POLYGON ((147.09960 -31.04462, 147.30960 -31.0...
4      POLYGON ((147.49960 -32.44461, 147.70960 -32.4...
                             ...                        
476    POLYGON ((150.29960 -34.24462, 150.50960 -34.2...
477    POLYGON ((150.29960 -34.04462, 150.50960 -34.0...
478    POLYGON ((150.29960 -34.04462, 150.50960 -34.0...
479    POLYGON ((150.29960 -33.84462, 150.50960 -33.8...
480    POLYGON ((150.29960 -33.84462, 150.50960 -33.8...
Name: geometry, Length: 481, dtype: geometry

Getting the bounds from a single polygon

In [11]:
test_bounds= sydney_aoi.geometry[0].bounds
test_bounds

(147.09959654580408,
 -31.244618412374592,
 147.30959654580408,
 -31.03461841237459)

# Sorting the dates

Converting standard date format into a datetime objectget_landsat_collection(date_start= , date_end= , bounds= test_bounds)

In [12]:
date_string_start = '02/08/21'
date_string_end = '03/09/21'
datetime_start= datetime.strptime(date_string_start, '%d/%m/%y')
datetime_end= datetime.strptime(date_string_end, '%d/%m/%y')
datetime_start

datetime.datetime(2021, 8, 2, 0, 0)

In [13]:
date_range = ('02/08/21', '03/09/21')

In [14]:
type(date_range)

tuple

In [15]:
#Test of datetime function from query notebook

#e.g. date_range = ('dd/mm/yy', 'dd/mm/yy')

def get_datetime(date_range):
    datetime_start= datetime.strptime(date_range[0], '%d/%m/%y')
    datetime_end= datetime.strptime(date_range[1], '%d/%m/%y')
    
    start_year = datetime_start.year
    end_year = datetime_end.year
    
    return datetime_start, datetime_end, start_year, end_year

dates = get_datetime(date_range)

print(dates)

(datetime.datetime(2021, 8, 2, 0, 0), datetime.datetime(2021, 9, 3, 0, 0), 2021, 2021)


In [16]:
type(dates[0])

datetime.datetime

In [17]:
print(f'Start datetime: {dates[0]}, end datetime: {dates[1]}')

Start datetime: 2021-08-02 00:00:00, end datetime: 2021-09-03 00:00:00


# Authenticating `ee`

First will have to run `earthengine authenticate` in the VM command line / terminal. 

**Issue:** Doesn't allow us to authenticate. We get an 'access blocked' message. We checked the GCP API information under APIs and Servicse > Enabled APIs & Services and there are no API credentials (client ID and key). We had a look into what the issue might be and found that we may need to enable 'Google Earth Engine API', but this did nothing. 

- Do we need client ID and key for the API to get access?
- Do we need to open the authentication link  in a VM browser? 
- Any other reasons why the authentication link doesn't work


**SOLVED** Cormac suggested to run the following

In [18]:
ee.Authenticate()

Enter verification code:  4/1ARtbsJrcfxy5g9NhnSzloYgWvcXhd46799pmS6l2tXFPkkZkvoUlMcHZTyo



Successfully saved authorization token.


Initialise earth engine (ee)

In [19]:
ee.Initialize()

# Getting the bounds and getting the images

In [20]:
type(sydney_aoi.geometry[0].bounds)

tuple

In [21]:

js = json.loads(sydney_aoi.to_json()) # Converting the aoi to a json and loading the json information, inludes coordinates,
                                        # id for each polygon. Within the dictionary we have a 'featureCollection'
    
bounds = ee.Geometry(ee.FeatureCollection(js).geometry()) # Getting the bounds of the polygons 

type(bounds)

ee.geometry.Geometry

In [22]:
landsat_imgs = get_landsat_collection(date_start=dates[0], date_end=dates[1] , bounds=bounds)

In [23]:
type(landsat_imgs)

ee.imagecollection.ImageCollection

# Checking the properties of the imageCollection

Checking the size of the image collection

NOTE: Find date range limits that won't time out

**Example** from https://developers.google.com/earth-engine/apidocs/ee-imagecollection-size

In [24]:
col = ee.ImageCollection('LANDSAT/LC08/C02/T1_TOA').filterBounds(ee.Geometry.Point(-90.70, 34.71)).filterDate('2020-01-01', '2021-01-01');

In [25]:
type(col)

ee.imagecollection.ImageCollection

[running the below code in this link gives the number of images](https://code.earthengine.google.com/820fcb75a86f86352d4ae1d62c4650fe)

**End Example**

After checking the function `_get_collection()` in the script, I noticed the variable `n_images` which seemed like 'number of images'. Using the same code from the script gives us: 

In [26]:
landsat_imgs.size().getInfo()

12

We can see the `CLOUD_SCORE` is contained in 

features: [] -> id: 'date', properties: {} -> CLOUD_SCORE

In [27]:
myDict = landsat_imgs.getInfo()

In [28]:
len(myDict['features'])

12

In [29]:
myDict['features'][0].get('id')

'2021-08-17'

In [30]:
myDict = landsat_imgs.getInfo()

In [31]:
myDict['features'][0].get('properties')['CLOUD_COVER']

3.48

In [32]:
cloudScore_landsat = np.array([])

for i in range(len(myDict['features'])):
    val = myDict['features'][i].get('properties')['CLOUD_COVER']
    cloudScore_landsat = np.append(cloudScore_landsat, val)
    print(i, val) 


print(f'Mean cloud score is {cloudScore_landsat.mean()}')

0 3.48
1 10.97
2 40.43
3 76.51
4 0.32
5 0.05
6 54.69
7 10.8
8 0.02
9 14.17
10 1.96
11 0.01
Mean cloud score is 17.784166666666668


Now we need to know what a cloud score looks like and if there are any other properties we want to consider from the imageCollection.

plot out on a spatial grid. 

Each patch may have multiple images with each of the cloud scores. 

geopandas table with whichever coordinates used. coords and cloud cover.
sleep statement after each point to see how the plot is being built. 

---

geoscape query and intersects. 

---

linking the keys for both satellites.

To access the information, the procses looks like this: `landsat_imgs.getInfo()['features'][my_index].get('properties')['my_feature']`. 

getInfo() is a dictionary, so we access the dictionary by calling one of the words in the dictionary being 'features'. This now becomes a list of the different images, so we can index into the image we want. We then use the get() function which again returns a dictionary we can access with one of the words in the dictionary to access the information. 

**Now getting the images for s2. Note, s2 has different properties to landsat**

In [33]:
get_s2_collection = ee_download.get_s2_collection

In [34]:
s2_imgs = get_s2_collection(date_start=dates[0], date_end=dates[1] , bounds=bounds)

In [35]:
s2_imgs.getInfo()['features'][0].get('properties')['CLOUD_COVERAGE_ASSESSMENT']

10.1582

Number of images

In [36]:
s2_imgs.size().getInfo()

26

In [37]:
myDict2 = s2_imgs.getInfo()

In [38]:
cloudScore_s2 = np.array([])

for i in range(s2_imgs.size().getInfo()):
    val = myDict2['features'][i].get('properties')['CLOUD_COVERAGE_ASSESSMENT']
    cloudScore_s2 = np.append(cloudScore_s2, val)
    print(i, val)
    
print(f'Mean cloud score is {cloudScore_s2.mean()}')

0 10.1582
1 18.4946
2 19.9737
3 57.2791
4 69.7648
5 0
6 1.937
7 6.9454
8 16.4855
9 0
10 0
11 29.3754
12 0
13 98.0081
14 0
15 0
16 90.0543
17 99.9021
18 0
19 51.3534
20 0
21 99.9965
22 99.9962
23 0.2658
24 0
25 0
Mean cloud score is 29.615003846153847


# Checking the shapefile data type 

In [39]:
import shutil

In [40]:
extract= shutil.unpack_archive('../data/Declarations_19722_Merge.zip')

In [41]:
test= gpd.read_file('../data/Declarations_unzip/Declarations_19722_Merge.shp')

In [42]:
type(test)

geopandas.geodataframe.GeoDataFrame

In [43]:
type(test['geometry'])

geopandas.geoseries.GeoSeries

In [44]:
type(test['geometry'][37])

shapely.geometry.multipolygon.MultiPolygon

# Putting useful information into a dataframe

* CLOUD_COVER
* polygon
* satellite image ID associated with the polygon
* date associated with the specific satellite image


We need to find out what other useful information is in the properties of the imageCollection 

Checking the number of polygons in `js` which comes from `sydney_aoi`

In [45]:
len(js['features'])

481

Can also be done with 

In [46]:
len(sydney_aoi)

481

Checking the `myDict['features'][0].get('properties')` dictionary, some interesting properties to explore are: 
* CLOUD_COVER
* cloud_probability
* DATE_ACQUIRED
* 'system:index' 

this is another date, one day after DATE_ACQUIRED.
* CLOUD_COVER_LAND
* COLLECTION_NUMBER

Checking the `myDict['features'][0].get('id')` dictionary has a date that is one day after DATE_acquired.


In [47]:
cloudScore_landsat

array([3.480e+00, 1.097e+01, 4.043e+01, 7.651e+01, 3.200e-01, 5.000e-02,
       5.469e+01, 1.080e+01, 2.000e-02, 1.417e+01, 1.960e+00, 1.000e-02])

# Extracting properties for individual polygons from original AOI

In [48]:
def get_imageCollections(bounds: ee.geometry.Geometry, dates: tuple): # returns imageCollection

    get_landsat_collection = ee_download.get_landsat_collection
    landsat_imgs = get_landsat_collection(date_start=dates[0], date_end=dates[1] , bounds= bounds)

    get_s2_collection = ee_download.get_s2_collection
    s2_imgs = get_s2_collection(date_start=dates[0], date_end=dates[1] , bounds= bounds)
    
    return landsat_imgs, s2_imgs

In [49]:
def get_datetime(date_range: tuple): # function returns a tuple containing datetime
    datetime_start= datetime.strptime(date_range[0], '%d/%m/%y')
    datetime_end= datetime.strptime(date_range[1], '%d/%m/%y')
    
    start_year = datetime_start.year
    end_year = datetime_end.year
    
    return datetime_start, datetime_end, start_year, end_year 

In [50]:
def read_aoi(my_gpd_aoi:str):
    my_aoi = gpd.read_file(my_gpd_aoi)
    
    aoi_codes = my_aoi['aoi_code']
    js = json.loads(my_aoi.to_json()) # convertin gthe aoi to a json and getting the 
    bounds = ee.Geometry(ee.FeatureCollection(js).geometry())
    
    return bounds, aoi_codes

In [51]:
def extract_properties(landsat_imgs, s2_imgs, init_bounds, aoi_code):
    
    coordinates = init_bounds['coordinates']
    
    landsat_dict = landsat_imgs.getInfo()
    s2_dict = s2_imgs.getInfo()
    
    n_landsat = len(landsat_dict['features'])
    print('Landsat images:', n_landsat)
    n_s2 = len(s2_dict['features'])
    print('s2 images:', n_s2)
    
    n_total = n_landsat + n_s2 
    
    properties_df = pd.DataFrame(index=np.arange(0, n_total), columns=['id', 'source', 'cloud_cover', 'percentage_valid_pixels', 'coordinates', 'aoi_code']) 
    
    
    for i in range(n_landsat):
        properties_df['id'][i] = landsat_dict['features'][i].get('id')
        properties_df['source'][i] = 'landsat'
        properties_df['cloud_cover'][i] = landsat_dict['features'][i].get('properties')['CLOUD_COVER']
        properties_df['percentage_valid_pixels'][i] = landsat_dict['features'][i].get('properties')['valids'] 
        properties_df['coordinates'][i] = coordinates
        properties_df['aoi_code'][i] = aoi_code
        #find area/co-ordinates associated with each image
        
    for i in range(n_s2):
        j = i + n_landsat
        
        properties_df['id'][j] = s2_dict['features'][i].get('id')
        properties_df['source'][j] = 's2'
        properties_df['cloud_cover'][j] = s2_dict['features'][i].get('properties')['CLOUD_COVERAGE_ASSESSMENT']
        properties_df['percentage_valid_pixels'][j] = s2_dict['features'][i].get('properties')['valids'] 
        properties_df['coordinates'][j] = coordinates
        properties_df['aoi_code'][j] = aoi_code
        #find area/co-ordinates associated with each image
    
    return properties_df

In [52]:
test_dates = ('02/08/21', '03/09/21')

test_datetime = get_datetime(test_dates)

In [53]:
sydney_aoi, sydney_aoi_codes = read_aoi('data/2022-07-21_aois.geojson')

sydney_aoi_codes

0        AOI01
1        AOI01
2        AOI02
3        AOI02
4      AOI0289
        ...   
476      AOI97
477      AOI98
478      AOI98
479      AOI99
480      AOI99
Name: aoi_code, Length: 481, dtype: object

In [54]:
sydney_aoi_all = sydney_aoi.getInfo()['coordinates']

sydney_aoi_all[0]

[[[147.09959654580408, -31.244618412374592],
  [147.30959654580408, -31.244618412374592],
  [147.30959654580408, -31.03461841237459],
  [147.09959654580408, -31.03461841237459],
  [147.09959654580408, -31.244618412374592]]]

In [55]:
len(sydney_aoi_all)

263

In [56]:
sydney_aoi_poly = ee.Geometry.Polygon(sydney_aoi_all[0])

type(sydney_aoi_poly)

ee.geometry.Geometry

In [57]:
len(sydney_aoi_all)

263

In [58]:
poly_landsat, poly_s2 = get_imageCollections(sydney_aoi_poly, test_datetime)

In [59]:
test_poly_df = extract_properties(poly_landsat, poly_s2, sydney_aoi_poly, sydney_aoi_codes[0])

Landsat images: 4
s2 images: 13


In [60]:
test_poly_df

id   source cloud_cover percentage_valid_pixels  \
0   2021-08-13  landsat        0.02                       1   
1   2021-08-29  landsat       14.17                       1   
2   2021-08-04  landsat        1.96                0.279306   
3   2021-08-20  landsat        0.01                0.276128   
4   2021-08-03       s2      3.0108                       1   
5   2021-08-05       s2      40.112                       1   
6   2021-08-08       s2      2.0378                       1   
7   2021-08-10       s2           0                       1   
8   2021-08-13       s2      0.0281                       1   
9   2021-08-15       s2      5.4465                       1   
10  2021-08-18       s2      0.2876                       1   
11  2021-08-20       s2      0.6071                       1   
12  2021-08-23       s2     33.5172                       1   
13  2021-08-25       s2      0.1036                       1   
14  2021-08-28       s2           0                       1   
15  2021-08-30       s2     97.7998                       1   
16  2021-09-02       s2           0                       1   

                                          coordinates aoi_code  
0   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
1   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
2   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
3   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
4   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
5   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
6   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
7   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
8   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
9   [[[147.09959654580408, -31.244618412374592], [...    AOI01  
10  [[[147.09959654580408, -31.244618412374592], [...    AOI01  
11  [[[147.09959654580408, -31.244618412374592], [...    AOI01  
12  [[[147.09959654580408, -31.244618412374592], [...    AOI01  
13  [[[147.09959654580408, -31.244618412374592], [...    AOI01  
14  [[[147.09959654580408, -31.244618412374592], [...    AOI01  
15  [[[147.09959654580408, -31.244618412374592], [...    AOI01  
16  [[[147.09959654580408, -31.244618412374592], [...    AOI01

In [61]:
test_poly_df['coordinates'][0]

[[[147.09959654580408, -31.244618412374592],
  [147.30959654580408, -31.244618412374592],
  [147.30959654580408, -31.03461841237459],
  [147.09959654580408, -31.03461841237459],
  [147.09959654580408, -31.244618412374592]]]

In [62]:
test_poly_df.shape

(17, 6)

**QUESTION** Previously we had 12 images for landsat and 26 images for s2, is the decrease we see above due to only selecting a single AOI from `sydney_aoi_codes`? 

`test_poly_df` Now has the properties from the iamges for a single polygon of an AOI, being from `sydney_aoi_all[0]`. We want to link all the AOI codes to each polygon. 

It seems like for each polygon we need to check each image and associate the AOI to that polygon.

# Getting the images for multiple polygons for Landsat

In [63]:
type(sydney_aoi_poly)

ee.geometry.Geometry

Getting the first 20 polygons instead of all 263 (long processing times when getting the imageCollection. If you run on all 263 polygons, takes about 10 minutes.

In [64]:
sydney_aoi, sydney_aoi_codes = read_aoi('data/2022-07-21_aois.geojson')

# sydney_aoi_codes

In [65]:
sydney_aoi_all = sydney_aoi.getInfo()['coordinates']

# sydney_aoi_all

In [66]:
sydney_aoi_poly = []
for i in range(len(sydney_aoi_all[0:20])):
    sydney_aoi_poly = np.append(sydney_aoi_poly, ee.Geometry.Polygon(sydney_aoi_all[i]))
    
type(sydney_aoi_poly)

numpy.ndarray

In [67]:
sydney_aoi_poly[0]['coordinates']

[[[147.09959654580408, -31.244618412374592],
  [147.30959654580408, -31.244618412374592],
  [147.30959654580408, -31.03461841237459],
  [147.09959654580408, -31.03461841237459],
  [147.09959654580408, -31.244618412374592]]]

Getting the first 20 codes to match the first 20 polygons

In [68]:
sydney_aoi_codes = sydney_aoi_codes[0:20]

In [69]:
len(sydney_aoi_poly)

20

In [70]:
def get_imageCollections(bounds: np.ndarray, dates: tuple): # returns imageCollection

    get_landsat_collection = ee_download.get_landsat_collection
    get_s2_collection = ee_download.get_s2_collection
    
    landsat_imgs = []
    s2_imgs = []
    
    for i in range(len(sydney_aoi_poly)):
        landsat_imgs = np.append(landsat_imgs, get_landsat_collection(date_start=dates[0], date_end=dates[1] , bounds= bounds[i]))
        s2_imgs = np.append(s2_imgs, get_s2_collection(date_start=dates[0], date_end=dates[1] , bounds= bounds[i]))

    return landsat_imgs, s2_imgs

In [71]:
poly_landsat, poly_s2 = get_imageCollections(sydney_aoi_poly, test_datetime)

In [72]:
len(poly_landsat[0].getInfo()['features'])

4

Checking how many images are in each of the firs 20 polygons

In [73]:
for i in range(len(poly_landsat)):
    print(len(poly_landsat[i].getInfo()['features']))    

4
4
2
2
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4
4


In [74]:
len(sydney_aoi_poly)

20

In [75]:
sydney_aoi_poly[0]['coordinates']

[[[147.09959654580408, -31.244618412374592],
  [147.30959654580408, -31.244618412374592],
  [147.30959654580408, -31.03461841237459],
  [147.09959654580408, -31.03461841237459],
  [147.09959654580408, -31.244618412374592]]]

In [76]:
landsat_coordinates = [[] for _ in range(len(poly_landsat))] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
n_landsat = []

for i in range(len(poly_landsat)):
    landsat_coordinates[i] = sydney_aoi_poly[i]['coordinates']

    n_landsat = np.append(n_landsat, len(poly_landsat[i].getInfo()['features'])) # Of the 20 indices, there will be the number of images in each. 


Trying to make a DF that isn't nested. Creating the DF comes later, first let's try to make lists that will be used to populate the DF. 

In [77]:
id_list = []
cloudCover_list = []
validPixels_list = []
# coordinates_list = []
# aoi_list = []
source_list = ['landsat']*np.int(n_landsat.sum()) # creating a list the size of the total number of images of all the polygons

for i in range(len(poly_landsat)): # Iterating through all the polygons
    
    for k in range(np.int(n_landsat[i])): # Iterating through all the images of each polygon
        
        id_list = np.append(id_list, poly_landsat[i].getInfo()['features'][k].get('id'))
        cloudCover_list = np.append(cloudCover_list, poly_landsat[i].getInfo()['features'][k].get('properties')['CLOUD_COVER'])
        validPixels_list = np.append(validPixels_list, poly_landsat[i].getInfo()['features'][k].get('properties')['valids'])
        # coordinates_list = landsat_coordinates[i]
        # aoi_list = sydney_aoi_codes[i]


Having issues with coordinates and aoi, commented them out and will work on them in a separate cell. 

In [78]:
len(landsat_coordinates)

20

In [79]:
landsat_coordinates[0]

[[[147.09959654580408, -31.244618412374592],
  [147.30959654580408, -31.244618412374592],
  [147.30959654580408, -31.03461841237459],
  [147.09959654580408, -31.03461841237459],
  [147.09959654580408, -31.244618412374592]]]

In [80]:
coordinates_list = [] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates

aoi_list = []

for i in range(len(poly_landsat)): # Iterating through all the polygons
    
    for k in range(len(poly_landsat[i].getInfo()['features'])): # Iterating through all the images of each polygon
        
        # print(k)
        
        aoi_list = np.append(aoi_list, sydney_aoi_codes[i])
        
        index_coord = landsat_coordinates[i]
        
        # print(index_coord)

        coordinates_list.extend(index_coord)

In [81]:
len(coordinates_list)

76

In [82]:
type(coordinates_list[0])

list

Managed to get the aoi list, but the coordinates list is having issues. For every polygon, the array of coordinates needs to be duplicated to the number of images that are in the polygon. 

In [83]:
landsat_df = gpd.GeoDataFrame({'id':id_list, 'cloud_cover':cloudCover_list, 'source': source_list, 'aoi_code': aoi_list, 'percentage_valid_pixels': validPixels_list, 'geometry':coordinates_list})
landsat_df.head(10)

id  cloud_cover   source aoi_code  percentage_valid_pixels  \
0  2021-08-13         0.02  landsat    AOI01                 1.000000   
1  2021-08-29        14.17  landsat    AOI01                 1.000000   
2  2021-08-04         1.96  landsat    AOI01                 0.279306   
3  2021-08-20         0.01  landsat    AOI01                 0.276128   
4  2021-08-13         0.02  landsat    AOI01                 1.000000   
5  2021-08-29        14.17  landsat    AOI01                 1.000000   
6  2021-08-04         1.96  landsat    AOI01                 0.521599   
7  2021-08-20         0.01  landsat    AOI01                 0.518510   
8  2021-08-13         2.36  landsat    AOI02                 1.000000   
9  2021-08-29        56.97  landsat    AOI02                 1.000000   

                                            geometry  
0  [[147.09959654580408, -31.244618412374592], [1...  
1  [[147.09959654580408, -31.244618412374592], [1...  
2  [[147.09959654580408, -31.244618412374592], [1...  
3  [[147.09959654580408, -31.244618412374592], [1...  
4  [[147.09959654580408, -31.04461841237459], [14...  
5  [[147.09959654580408, -31.04461841237459], [14...  
6  [[147.09959654580408, -31.04461841237459], [14...  
7  [[147.09959654580408, -31.04461841237459], [14...  
8  [[147.4996031608572, -32.44460560526437], [147...  
9  [[147.4996031608572, -32.44460560526437], [147...

#  Visualisation for Landsat

In [84]:
from shapely.geometry import Polygon

landsat_df['geometry'] = landsat_df['geometry'].apply(Polygon)

In [85]:
landsat_df

id  cloud_cover   source aoi_code  percentage_valid_pixels  \
0   2021-08-13         0.02  landsat    AOI01                 1.000000   
1   2021-08-29        14.17  landsat    AOI01                 1.000000   
2   2021-08-04         1.96  landsat    AOI01                 0.279306   
3   2021-08-20         0.01  landsat    AOI01                 0.276128   
4   2021-08-13         0.02  landsat    AOI01                 1.000000   
..         ...          ...      ...      ...                      ...   
71  2021-08-29        56.97  landsat  AOI0395                 0.705777   
72  2021-08-06        54.69  landsat  AOI0396                 1.000000   
73  2021-08-22        10.80  landsat  AOI0396                 1.000000   
74  2021-08-13         2.36  landsat  AOI0396                 0.932296   
75  2021-08-29        56.97  landsat  AOI0396                 0.938963   

                                             geometry  
0   POLYGON ((147.09960 -31.24462, 147.30960 -31.2...  
1   POLYGON ((147.09960 -31.24462, 147.30960 -31.2...  
2   POLYGON ((147.09960 -31.24462, 147.30960 -31.2...  
3   POLYGON ((147.09960 -31.24462, 147.30960 -31.2...  
4   POLYGON ((147.09960 -31.04462, 147.30960 -31.0...  
..                                                ...  
71  POLYGON ((148.29960 -32.24461, 148.50960 -32.2...  
72  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  
73  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  
74  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  
75  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  

[76 rows x 6 columns]

In [86]:
landsat_df = landsat_df.set_crs('EPSG:4326')

landsat_df.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

In [87]:
!pip install folium matplotlib mapclassify

In [88]:
landsat_df.explore()

# Getting the images for multiple polygons for S2

Following section has been made raw cells as they will be used in the next section

This should match the loop above

Checking the s2 image properties

Choosing s2 image properties from [here](https://developers.google.com/earth-engine/datasets/catalog/COPERNICUS_S2_SR#image-properties) and from above: 

* CLOUDY_PIXEL_PERCENTAGE (link)
* valids (above)
* id (not a property)


# Putting the loops together for both satellites

Cells have been converted to raw cell as they have been merged in the next section. 

Takes about 8 minutes to run 

takes about 12 minutes to run 

# Putting the two loops together

Can't put them in the same loops due to variable names becoming mixed up and loop lengths being different. 

In [89]:
def read_aoi(my_gpd_aoi:str):
    my_aoi = gpd.read_file(my_gpd_aoi)
    
    aoi_codes = my_aoi['aoi_code']
    js = json.loads(my_aoi.to_json()) # convertin gthe aoi to a json and getting the 
    
    total_bounds = ee.Geometry(ee.FeatureCollection(js).geometry())
    total_bounds_coordinates = total_bounds.getInfo()['coordinates']
    
    n_polygons = len(total_bounds_coordinates)
    
    print('Number of polygons in AOI:', n_polygons)
    
    return total_bounds, aoi_codes, n_polygons


In [90]:
# Read all the AOI.
sydney_aoi, sydney_aoi_codes, sydney_polygons = read_aoi('data/2022-07-21_aois.geojson')


Number of polygons in AOI: 263


In [91]:
# Get the coordinates for each AOI.
sydney_aoi_all = sydney_aoi.getInfo()['coordinates']

# Reducing the number of AOIs to 20. 
sydney_aoi_codes = sydney_aoi_codes[0:20]

# Updating a list with the first 20 coordinates due to processing times. 
sydney_aoi_poly = []
for i in range(len(sydney_aoi_all[0:20])):
    
    sydney_aoi_poly = np.append(sydney_aoi_poly, ee.Geometry.Polygon(sydney_aoi_all[i]))

Putting the above process into a function that returns a list of the bounds of the polygons, in which you can specify how many polygons you want to include

In [92]:
def get_polygon_bounds(total_bounds, aoi_codes, n_polygons):
    # Get the coordinates for each AOI.
    all_polygon_bounds = total_bounds.getInfo()['coordinates']

    # Reducing the number of AOIs to n_polygons 
    polygon_aoi_codes = aoi_codes[0:n_polygons]

    # Updating a list with the first 20 coordinates due to processing times. 
    polygon_bounds = []
    for i in range(len(all_polygon_bounds[0:n_polygons])):
      
       polygon_bounds = np.append(polygon_bounds, ee.Geometry.Polygon(all_polygon_bounds[i]))
                       
    return polygon_bounds, polygon_aoi_codes

In [93]:
sydney_20_polys, sydney_20_aois = get_polygon_bounds(sydney_aoi, sydney_aoi_codes, 20)

In [94]:
def get_imageCollections(polygon_bounds: np.ndarray, dates: tuple): #bounds_list refers to a list of geometries/polygons extracted above from the aoi

    get_landsat_collection = ee_download.get_landsat_collection
    get_s2_collection = ee_download.get_s2_collection
    
    landsat_imgs = []
    s2_imgs = []
    
    for i in range(len(polygon_bounds)):
        landsat_imgs = np.append(landsat_imgs, get_landsat_collection(date_start=dates[0], date_end=dates[1] , bounds= polygon_bounds[i]))
        s2_imgs = np.append(s2_imgs, get_s2_collection(date_start=dates[0], date_end=dates[1] , bounds= polygon_bounds[i]))

    return landsat_imgs, s2_imgs

In [95]:
dates = ('02/08/21', '03/09/21')

dateTime = get_datetime(dates)

In [96]:
poly_landsat, poly_s2 = get_imageCollections(sydney_20_polys, dateTime)

In [97]:
# --------- Landsat ---------
landsat_coordinates = [[] for _ in range(len(poly_landsat))] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
n_landsat = []

# Loop 1: Getting the number of images for each AOI, returns an array where each index is an 
# AOI and each value in the index is the number of images of that AOI. 
for i in range(len(poly_landsat)):
    
    # Get the coordinates for each AOI
    landsat_coordinates[i] = sydney_20_polys[i]['coordinates']
    
    # Of the 20 indices, there will be the number of images in each. 
    n_landsat = np.append(n_landsat, len(poly_landsat[i].getInfo()['features']))

# End loop 1

id_list = []
cloudCover_list = []
validPixels_list = []
source_list = ['landsat']*np.int(n_landsat.sum()) # creating a list the size of the total number of images of all the polygons
coordinates_list = [] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
aoi_list = []

# Loop 2: Updating property lists that will be used to populate a dataframe.
for i in range(len(poly_landsat)): # Iterating through all the polygons
    
    for k in range(len(poly_landsat[i].getInfo()['features'])): # Iterating through all the images of each polygon
        
        id_list = np.append(id_list, poly_landsat[i].getInfo()['features'][k].get('id'))
        cloudCover_list = np.append(cloudCover_list, poly_landsat[i].getInfo()['features'][k].get('properties')['CLOUD_COVER'])
        validPixels_list = np.append(validPixels_list, poly_landsat[i].getInfo()['features'][k].get('properties')['valids'])
        aoi_list = np.append(aoi_list, sydney_aoi_codes[i])
        index_coord = landsat_coordinates[i]
        coordinates_list.extend(index_coord)      
# End loop 2

# Populate the dataframe
landsat_df = gpd.GeoDataFrame({'id':id_list, 'cloud_cover':cloudCover_list, 'source': source_list, 'aoi_code': aoi_list, 'percentage_valid_pixels': validPixels_list, 'geometry':coordinates_list})
landsat_df.head(10)
print('Number of landsat images:', len(landsat_df))

# --------- S2 ---------
s2_coordinates = [[] for _ in range(len(poly_s2))] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
n_s2 = []
# Loop 1
for i in range(len(poly_s2)): # Iterating through all the polygons
    
    s2_coordinates[i] = sydney_20_polys[i]['coordinates']

    n_s2 = np.append(n_s2, len(poly_s2[i].getInfo()['features'])) # Of the 20 indices, there will be the number of images in each. 
# End loop 1

id_list = []
cloudCover_list = []
validPixels_list = []
source_list = ['s2']*np.int(n_s2.sum()) # creating a list the size of the total number of images of all the polygons
coordinates_list = [] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
aoi_list = []

# Loop 2
for i in range(len(poly_s2)): # Iterating through all the polygons
        
    for k in range(np.int(n_s2[i])): # Iterating through all the images of each polygon
        
        id_list = np.append(id_list, poly_s2[i].getInfo()['features'][k].get('id'))
        cloudCover_list = np.append(cloudCover_list, poly_s2[i].getInfo()['features'][k].get('properties')['CLOUDY_PIXEL_PERCENTAGE'])
        validPixels_list = np.append(validPixels_list, poly_s2[i].getInfo()['features'][k].get('properties')['valids'])
        aoi_list = np.append(aoi_list, sydney_aoi_codes[i])
        index_coord = s2_coordinates[i]
        coordinates_list.extend(index_coord)      
# End loop 2

# Populate the dataframe
s2_df = gpd.GeoDataFrame({'id':id_list, 'cloud_cover':cloudCover_list, 'source': source_list, 'aoi_code': aoi_list, 'percentage_valid_pixels': validPixels_list, 'geometry':coordinates_list})
s2_df.head(10)
print('Number of s2 images:', len(s2_df))

Number of landsat images: 76
Number of s2 images: 148


In [98]:
landsat_df.head(10)

id  cloud_cover   source aoi_code  percentage_valid_pixels  \
0  2021-08-13         0.02  landsat    AOI01                 1.000000   
1  2021-08-29        14.17  landsat    AOI01                 1.000000   
2  2021-08-04         1.96  landsat    AOI01                 0.279306   
3  2021-08-20         0.01  landsat    AOI01                 0.276128   
4  2021-08-13         0.02  landsat    AOI01                 1.000000   
5  2021-08-29        14.17  landsat    AOI01                 1.000000   
6  2021-08-04         1.96  landsat    AOI01                 0.521599   
7  2021-08-20         0.01  landsat    AOI01                 0.518510   
8  2021-08-13         2.36  landsat    AOI02                 1.000000   
9  2021-08-29        56.97  landsat    AOI02                 1.000000   

                                            geometry  
0  [[147.09959654580408, -31.244618412374592], [1...  
1  [[147.09959654580408, -31.244618412374592], [1...  
2  [[147.09959654580408, -31.244618412374592], [1...  
3  [[147.09959654580408, -31.244618412374592], [1...  
4  [[147.09959654580408, -31.04461841237459], [14...  
5  [[147.09959654580408, -31.04461841237459], [14...  
6  [[147.09959654580408, -31.04461841237459], [14...  
7  [[147.09959654580408, -31.04461841237459], [14...  
8  [[147.4996031608572, -32.44460560526437], [147...  
9  [[147.4996031608572, -32.44460560526437], [147...

In [99]:
all_df = pd.concat([landsat_df, s2_df])
all_df

id  cloud_cover   source aoi_code  percentage_valid_pixels  \
0    2021-08-13       0.0200  landsat    AOI01                 1.000000   
1    2021-08-29      14.1700  landsat    AOI01                 1.000000   
2    2021-08-04       1.9600  landsat    AOI01                 0.279306   
3    2021-08-20       0.0100  landsat    AOI01                 0.276128   
4    2021-08-13       0.0200  landsat    AOI01                 1.000000   
..          ...          ...      ...      ...                      ...   
143  2021-08-10       0.1973       s2  AOI0396                 1.000000   
144  2021-08-15       0.0231       s2  AOI0396                 1.000000   
145  2021-08-20       0.0000       s2  AOI0396                 1.000000   
146  2021-08-25       0.0179       s2  AOI0396                 1.000000   
147  2021-08-30      99.9943       s2  AOI0396                 1.000000   

                                              geometry  
0    [[147.09959654580408, -31.244618412374592], [1...  
1    [[147.09959654580408, -31.244618412374592], [1...  
2    [[147.09959654580408, -31.244618412374592], [1...  
3    [[147.09959654580408, -31.244618412374592], [1...  
4    [[147.09959654580408, -31.04461841237459], [14...  
..                                                 ...  
143  [[148.29960316085715, -32.044605605264366], [1...  
144  [[148.29960316085715, -32.044605605264366], [1...  
145  [[148.29960316085715, -32.044605605264366], [1...  
146  [[148.29960316085715, -32.044605605264366], [1...  
147  [[148.29960316085715, -32.044605605264366], [1...  

[224 rows x 6 columns]

In [100]:
all_df['geometry'] = all_df['geometry'].apply(Polygon)

In [101]:
all_df.crs = 'EPSG:4326'

In [102]:
all_df_sorted = all_df.sort_values(by = ['aoi_code', 'id'])

all_df_sorted = all_df_sorted.reset_index()

all_df_sorted

index          id  cloud_cover   source aoi_code  \
0        0  2021-08-03       3.0108       s2    AOI01   
1       13  2021-08-03       3.0108       s2    AOI01   
2        2  2021-08-04       1.9600  landsat    AOI01   
3        6  2021-08-04       1.9600  landsat    AOI01   
4        1  2021-08-05      40.1120       s2    AOI01   
..     ...         ...          ...      ...      ...   
219    145  2021-08-20       0.0000       s2  AOI0396   
220     73  2021-08-22      10.8000  landsat  AOI0396   
221    146  2021-08-25       0.0179       s2  AOI0396   
222     75  2021-08-29      56.9700  landsat  AOI0396   
223    147  2021-08-30      99.9943       s2  AOI0396   

     percentage_valid_pixels  \
0                   1.000000   
1                   1.000000   
2                   0.279306   
3                   0.521599   
4                   1.000000   
..                       ...   
219                 1.000000   
220                 1.000000   
221                 1.000000   
222                 0.938963   
223                 1.000000   

                                              geometry  
0    POLYGON ((147.09960 -31.24462, 147.30960 -31.2...  
1    POLYGON ((147.09960 -31.04462, 147.30960 -31.0...  
2    POLYGON ((147.09960 -31.24462, 147.30960 -31.2...  
3    POLYGON ((147.09960 -31.04462, 147.30960 -31.0...  
4    POLYGON ((147.09960 -31.24462, 147.30960 -31.2...  
..                                                 ...  
219  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  
220  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  
221  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  
222  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  
223  POLYGON ((148.29960 -32.04461, 148.50960 -32.0...  

[224 rows x 7 columns]

Is it worth doing this sorting within the main function or would it be better to produce the dataframe as is? 

In [103]:
all_df_sorted.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

Looking a the the visualisation it seems thathe .explore() automatically plots the images on top of each other and then only displays the stats for the top/last one in the hover menu. 

In [104]:
all_df_sorted.explore()

# Dataframe Function Testing


In [105]:
def properties_dataframe(landsat_imgs: ee.ImageCollection, s2_imgs: ee.ImageCollection, polygon_bounds, polygon_aoi_codes):
    # --------- Landsat ---------
    landsat_coordinates = [[] for _ in range(len(landsat_imgs))] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
    n_landsat = []

    # Loop 1: Getting the number of images for each AOI, returns an array where each index is an 
    # AOI and each value in the index is the number of images of that AOI. 
    for i in range(len(landsat_imgs)):

        # Get the coordinates for each AOI
        landsat_coordinates[i] = polygon_bounds[i]['coordinates']

        # Of the 20 indices, there will be the number of images in each. 
        n_landsat = np.append(n_landsat, len(landsat_imgs[i].getInfo()['features']))

    # End loop 1

    id_list = []
    cloudCover_list = []
    validPixels_list = []
    source_list = ['landsat']*np.int(n_landsat.sum()) # creating a list the size of the total number of images of all the polygons
    coordinates_list = [] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
    aoi_list = []

    # Loop 2: Updating property lists that will be used to populate a dataframe.
    for i in range(len(landsat_imgs)): # Iterating through all the polygons

        for k in range(np.int(n_landsat[i])): # Iterating through all the images of each polygon

            id_list = np.append(id_list, landsat_imgs[i].getInfo()['features'][k].get('id'))
            cloudCover_list = np.append(cloudCover_list, landsat_imgs[i].getInfo()['features'][k].get('properties')['CLOUD_COVER'])
            validPixels_list = np.append(validPixels_list, landsat_imgs[i].getInfo()['features'][k].get('properties')['valids'])
            aoi_list = np.append(aoi_list, polygon_aoi_codes[i])
            index_coord = landsat_coordinates[i]
            coordinates_list.extend(index_coord)      
    # End loop 2

    # Populate the dataframe
    landsat_df = gpd.GeoDataFrame({'id':id_list, 'cloud_cover':cloudCover_list, 'source': source_list, 'aoi_code': aoi_list, 'percentage_valid_pixels': validPixels_list, 'geometry':coordinates_list})
    landsat_df.head(10)
    print('Number of landsat images:', len(landsat_df))

    # --------- S2 ---------
    s2_coordinates = [[] for _ in range(len(s2_imgs))] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
    n_s2 = []
    # Loop 1
    for i in range(len(s2_imgs)): # Iterating through all the polygons

        s2_coordinates[i] = polygon_bounds[i]['coordinates']

        n_s2 = np.append(n_s2, len(s2_imgs[i].getInfo()['features'])) # Of the 20 indices, there will be the number of images in each. 
    # End loop 1

    id_list = []
    cloudCover_list = []
    validPixels_list = []
    source_list = ['s2']*np.int(n_s2.sum()) # creating a list the size of the total number of images of all the polygons
    coordinates_list = [] # creating number of lists to match the number of polygons so that each polygon will have it's coordinates
    aoi_list = []

    # Loop 2
    for i in range(len(s2_imgs)): # Iterating through all the polygons

        for k in range(np.int(n_s2[i])): # Iterating through all the images of each polygon

            id_list = np.append(id_list, s2_imgs[i].getInfo()['features'][k].get('id'))
            cloudCover_list = np.append(cloudCover_list, s2_imgs[i].getInfo()['features'][k].get('properties')['CLOUDY_PIXEL_PERCENTAGE'])
            validPixels_list = np.append(validPixels_list, s2_imgs[i].getInfo()['features'][k].get('properties')['valids'])
            aoi_list = np.append(aoi_list, polygon_aoi_codes[i])
            index_coord = s2_coordinates[i]
            coordinates_list.extend(index_coord)      
    # End loop 2

    # Populate the dataframe
    s2_df = gpd.GeoDataFrame({'id':id_list, 'cloud_cover':cloudCover_list, 'source': source_list, 'aoi_code': aoi_list, 'percentage_valid_pixels': validPixels_list, 'geometry':coordinates_list})
    s2_df.head(10)
    print('Number of s2 images:', len(s2_df))
    
    # Combining into one dataframe, assigning crs and formatting geometry column for plotting
    
    all_df = pd.concat([landsat_df, s2_df])
    
    all_df.crs = 'EPSG:4326'
    
    all_df['geometry'] = all_df['geometry'].apply(Polygon)
    
    print('Total number of images:', len(all_df))
          
    return all_df

## 5 Polygon Test

In [106]:
sydney_5_poly, sydney_5_aois = get_polygon_bounds(sydney_aoi, sydney_aoi_codes, 5)


In [108]:
poly_5_landsat, poly_5_s2 = get_imageCollections(sydney_5_poly, dateTime)

In [109]:
len(poly_5_landsat)

5

In [110]:
len(sydney_5_aois)

5

In [111]:
all5_df = properties_dataframe(poly_5_landsat, poly_5_s2, sydney_5_poly, sydney_5_aois)

Number of landsat images: 16
Number of s2 images: 58
Total number of images: 74


runtime of the above function over 5 polygons and 74 images: 12 minutes

In [112]:
all5_df.explore()

# To attempt:

from meeting 13/09/22
* summary information over the squares gives a timeseries plot (out of scope)
* blue squares to be shaded to a property
* explore function documentation to color each square by a certain property
* explore plotting functions
* check additional properties we can use. 